overlord/snapstate/component.go

```diff
@@ -0,0 +1,227 @@
+	// overkill and needs to be revisited).
+	if err := checkChangeConflictIgnoringOneChange(st, snapsup.InstanceName(),
+		snapst, fromChange); err != nil {
+		return nil, err
+	}
+
+	// check if we already have the revision in the snaps folder (alters tasks)
+	revisionIsPresent := snapst.IsComponentRevInstalled(snapSi, compSi) == true
+	revisionStr := fmt.Sprintf(" (%s)", compSi.Revision)
+
+	var prepare, prev *state.Task
+	// if we have a local revision here we go back to that
+	if path != "" || revisionIsPresent {
+		prepare = st.NewTask("prepare-component",
+			fmt.Sprintf(i18n.G("Prepare component %q%s"),
+				path, revisionStr))
+	} else {
+		// TODO implement download-component
+		return nil, fmt.Errorf("download-component not implemented yet")
+	}
+	prepare.Set("component-setup", compSetup)
+
+	tasks := []*state.Task{prepare}
+	prev = prepare
+
+	addTask := func(t *state.Task) {
+		t.Set("component-setup-task", prepare.ID())
+		t.WaitFor(prev)
+		tasks = append(tasks, t)
+	}
+
+	// TODO task to fetch and check assertions for component if from store
+	// (equivalent to "validate-snap")
+
+	// Task that copies the file and creates mount units
+	if !revisionIsPresent {
+		mount := st.NewTask("mount-component",
+			fmt.Sprintf(i18n.G("Mount component %q%s"),
+				compSi.Component, revisionStr))
+		addTask(mount)
+		prev = mount
+	} else {
+		if removeComponentPath {
+			// If the revision is local, we will not need the
+			// temporary snap. This can happen when e.g.
+			// side-loading a local revision again. The path is
+			// only needed in the "mount-snap" handler and that is
+			// skipped for local revisions.
+			if err := os.Remove(path); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// TODO hooks for components
+
+	// We might be replacing a component if a local install, otherwise
+	// this is not really possible.
+	compInstalled := snapst.IsComponentInstalled(compSi.Component)
+	if compInstalled {
+		unlink := st.NewTask("unlink-current-component", fmt.Sprintf(i18n.G(
+			"Make current revision for component %q unavailable"),
+			compSi.Component))
+		addTask(unlink)
+		prev = unlink
+	}
+
+	// finalize (sets SnapState)
+	linkSnap := st.NewTask("link-component",
+		fmt.Sprintf(i18n.G("Make component %q%s available to the system"),
+			compSi.Component, revisionStr))
+	addTask(linkSnap)
+	prev = linkSnap
+
+	installSet := state.NewTaskSet(tasks...)
+	installSet.MarkEdge(prepare, BeginEdge)
+	installSet.MarkEdge(linkSnap, MaybeRebootEdge)
+
+	// TODO if snap is being installed from the store, then the last task
+	// before any system modifications are done will be validate-component,
+	// otherwise it will be prepare-component. Change when
+	// validate-component is implemented.
+	installSet.MarkEdge(prepare, LastBeforeLocalModificationsEdge)
+
+	if err := SetEssentialSnapsRestartBoundaries(st, nil,
+		[]*state.TaskSet{installSet}); err != nil {
+		return nil, err
+	}
```

Comment:
this bits feels a bit premature, maybe we should leave a question mark todo instead?

Reply:
Right, I've replaced with a TODO